src/latte_kernel/nbe.cljc

```diff
@@ -169,3 +170,28 @@
  ([t] (norm defenv/empty-env [] t))
  ([def-env ctx t]
   (quotation (evaluation def-env ctx t))))
+
+(defn- readable-quotation-
+  [taken t]
+  (let [quot (partial readable-quotation- taken)]
+    (cond
+      ;; we fetch the original name of the var and make a fresh one out of it
+      ;; a variable without metadata is guaranteed to be a free variable
+      (stx/variable? t)
+      (if-let [name (::name (meta t))]
+        (stx/mk-fresh name taken)
+        t)
+
+      (stx/app? t)
+      (mapv quot t)
+
+      (or (stx/binder? t) (stx/ref? t) (stx/ascription? t))
```

Comment:
if it's a binder I think the binder name should be added to the `taken` set

Reply:
In the term
```clojure
(let [x1 (with-meta '_1 {::name 'x})]
  (list 'λ [x1 '✳] [x1 'x]))
```
We need both `x1` to be changed to `x'`. For the first one we do `(mk-fresh 'x #{'x})` which returns `x'`. Since we want exactly the same result for the second `x1`, we call `mk-fresh` with the same arguments.
If we added `x'` to `taken`, we would obtain `(λ [x' ✳] [x'' x])` which is false.

Reply:
I agree there is a problem when we have multiple lambdas though
I'm on it

Reply:
Fixing this problem somehow fixed everything. Will update PR